s2s/time.py

```diff
@@ -170,6 +173,48 @@ def map_years(
 
         return index
 
+    def map_to_data(self, input_data: Union[pd.Series, pd.DataFrame, xr.Dataset, xr.DataArray],
+        flat: bool = False
+        ) -> pd.DataFrame:
+        """Map the calendar to input data period.
+        
+        Get datetime range from input data and generate corresponding interval index. This method
+        guarentees that the generated interval (calendar) indices would be covered by the input
+        data.
+        Args:
+            input_data: Input data for datetime mapping. Its index must be pandas.DatetimeIndex.
+            flat: Same as the argument in ``map_years``.
+        Returns:
+            Pandas DataFrame filled with Intervals of the calendar's frequency.
+            (also see ``map_year`` and ``map_years``)
+        """
+        # check the datetime order of input data
+        first_timestamp = input_data.index.min()
+        last_timestamp = input_data.index.max()
+
+        map_last_year = last_timestamp.year
+        # ensure that the input data could always cover the advent calendar
+        map_first_year = first_timestamp.year + 1
```

Comment:
Consider e.g. anchor date 31 Dec and freq 7 days. Then adding that year is not necessary, and would ideally be avoided. Can you check it in a similar way as for the last date?

Reply:
Thanks for spotting this! I added a "if" check for the `map_first_year`:

```py
# ensure that the input data could always cover the advent calendar
if self.map_year(first_timestamp.year).iloc[-1].left > first_timestamp:
    map_first_year = first_timestamp.year
else:
    map_first_year = first_timestamp.year + 1
```

Now the case you mentioned here is covered by the function. I put it in the test as well:
```py
    def test_map_to_data_single_year_coverage(self):
        # test the single year coverage
        cal = AdventCalendar(anchor_date=(12, 31), freq='180d')
        # multiple years covered
        time_index = pd.date_range('20210101', '20211231', freq='7d')
        test_data = np.random.random(len(time_index))
        timeseries = pd.Series(test_data, index=time_index)
        year = cal.map_to_data(timeseries)

        expected = np.array(
            [
                interval("2021-07-04", "2021-12-31"),
                interval("2021-01-05", "2021-07-04"),
            ]
        )

        assert np.array_equal(year, expected)
```